cpp/tests/experimental/graph_test.cpp

```diff
@@ -0,0 +1,246 @@
+        std::sort(cugraph_pairs.begin(), cugraph_pairs.end());
+        ASSERT_TRUE(
+          std::equal(reference_pairs.begin(), reference_pairs.end(), cugraph_pairs.begin()))
+          << "Graph compressed sparse format indices & weights for vertex " << i
+          << " do not match with the reference values.";
+      } else {
+        std::vector<vertex_t> reference_indices(h_reference_indices.begin() + start,
+                                                h_reference_indices.begin() + (start + degree));
+        std::vector<vertex_t> cugraph_indices(h_cugraph_indices.begin() + start,
+                                              h_cugraph_indices.begin() + (start + degree));
+        std::sort(reference_indices.begin(), reference_indices.end());
+        std::sort(cugraph_indices.begin(), cugraph_indices.end());
+        ASSERT_TRUE(
+          std::equal(reference_indices.begin(), reference_indices.end(), cugraph_indices.begin()))
+          << "Graph compressed sparse format indices for vertex " << i
+          << " do not match with the reference values.";
+      }
+    }
+  }
+};
+
+// FIXME: add tests for type combinations
+TEST_P(Tests_Graph, CheckStoreTransposedFalse)
+{
+  run_current_test<uint32_t, uint32_t, float, false>(GetParam());
+  run_current_test<uint32_t, uint64_t, float, false>(GetParam());
+  run_current_test<uint64_t, uint64_t, float, false>(GetParam());
+  run_current_test<uint32_t, uint32_t, double, false>(GetParam());
+  run_current_test<uint32_t, uint64_t, double, false>(GetParam());
+  run_current_test<uint64_t, uint64_t, double, false>(GetParam());
+}
+
+// FIXME: add tests for type combinations
+TEST_P(Tests_Graph, CheckStoreTransposedTrue)
+{
+  run_current_test<uint32_t, uint32_t, float, true>(GetParam());
+  run_current_test<uint32_t, uint64_t, float, true>(GetParam());
+  run_current_test<uint64_t, uint64_t, float, true>(GetParam());
+  run_current_test<uint32_t, uint32_t, double, true>(GetParam());
+  run_current_test<uint32_t, uint64_t, double, true>(GetParam());
+  run_current_test<uint64_t, uint64_t, double, true>(GetParam());
+}
+
+INSTANTIATE_TEST_CASE_P(simple_test,
+                        Tests_Graph,
+                        ::testing::Values(Graph_Usecase("test/datasets/karate.mtx", false),
+                                          Graph_Usecase("test/datasets/karate.mtx", true),
+                                          Graph_Usecase("test/datasets/web-Google.mtx", false),
+                                          Graph_Usecase("test/datasets/web-Google.mtx", true),
+                                          Graph_Usecase("test/datasets/ljournal-2008.mtx", false),
+                                          Graph_Usecase("test/datasets/ljournal-2008.mtx", true),
+                                          Graph_Usecase("test/datasets/webbase-1M.mtx", false),
+                                          Graph_Usecase("test/datasets/webbase-1M.mtx", true)));
+
```

Comment:
We also probably want to test :
- `multi_GPU = true/false` paths 
- `multi_GPU = True` but there's only one GPU (this is what will run on CI with the current setup)
This could be tested at the python level also

We also need to think about `multi_GPU = True` and Comm is set to use a subset of GPUs.

Reply:
For true multi-GPU C++ tests, AFAIK, we don't have an infrastructure for this.

https://github.com/rapidsai/raft/pull/63 is recently merged to RAFT, and AFAIK, @rlratzel is working on C++ infrastructure for multi-GPU testing based on the above PR.

I can take a look for the multi-GPU with only one GPU case.

I don't think "We also need to think about multi_GPU = True and Comm is set to use a subset of GPUs." really matters. This code just works with the GPUs in the comm in the passed handle. No difference whether all the GPUs in the system is used or only a subset of the GPUs are used.

Reply:
> I can take a look for the multi-GPU with only one GPU case.

CI will currently run in this context if you just set multi GPU to true

> This code just works with the GPUs in the comm in the passed handle. 

Agreed from a  theoretical perspective. In practice, we got surprises in the past. This is can be added later though.

Reply:
Thought about adding MG tests, but I think this needs more discussion.

Are we going to have two separate gtests executable files or have one?

I assume SG version will be run as gtests/EXECUTABLE_NAME and MG version should be run as mpiexec -n #_gpus gtests/EXECUTABLE_NAME.

Mixing SG & MG tests in one executable may still work (but mpiexec -n #_gpus will run the same SG tests in every GPU) but can be wasteful. I think we need to decide our policy on running SG/MG C++ tests first before jumping to this (and are these C++ tests part of our CI/CD and benchmarking?)